### PR TITLE
QG Truth-twin / Observation time series generation

### LINKED ISSUE #94: QG Model Observation Vector / State Vector Relative Indexing
I'm working on a simulation built around the idea of the demo.py in the QG model where I'd like to visualize the time series of the truth twin and the associated observations of the truth twin from Sakov2008, plotted side-by-side on the same relative state vector 2D grid.  Extracting the relative indices during run-time in the HMM.simulate seems like a reasonable approach, as these indices are defined with a random offset.  However, before I do anything too invasive I wanted to check about the proposed approach.  The end goal is just to get an intuitive visualization of the truth-twin and the noisy, limited observations of the truth twin as the observation operator sweeps the 2D grid in real time, to better understand both processes.

### PR DESCRIPTION
**Pull request by cgrudz**
This is a pretty rough draft from our discussion in  `resolves #94`, there are probably smarter ways to do this so I'll definitely take suggestions, but this gets the main idea that I want to share with a collaborator here interested in working with learning problems on this model.

### CODE REVIEW

File: examples/generate_QG_time_series.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+## generate truth-twin and observation time series as per standard configuration
+## of Sakov2008, save the double time series
+xx, yy = HMM.simulate()
+
+# save data for easy re-use and debugging
+np.savez("QG_truth_obs_time_series.npz", xx=xx, yy=yy)
+
+# these lines can be used in place of above steps after first run,
+# comment out above
+tmp = np.load("QG_truth_obs_time_series.npz")
```
Comment (patricknraanes) [inline examples/generate_QG_time_series.py:47]: `with` block is nicer than having to `close` the file later

File: examples/generate_QG_time_series.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
+"""Truth-twin and observation time series for QG (quasi-geostrophic) model."""
```
Comment (patricknraanes) [inline examples/generate_QG_time_series.py:1]: I don't see the need for the "twin" wording in here (and elsewhere below). It's simply a forward simulation experiment. Also, even in DA settings, the wording used elsewhere in DAPPER is "synthetic truth" (with "twin" only sometimes put into parantheses or similar). Yes, it's my preference, but I stand by it, and as mentioned I even claim it's DAPPER "convention"

File: examples/generate_QG_time_series.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+# comment out above
+tmp = np.load("QG_truth_obs_time_series.npz")
+xx = tmp['xx'][1:,]
+yy = tmp['yy']
+tmp.close()
+
+# generate observations on the same gridding as the state vector
+
+# initialize the storage on the parent state dimension
+yy_xx = np.empty(np.shape(xx))
```
Comment (patricknraanes) [inline examples/generate_QG_time_series.py:55]: probs use `np.empty_like`

File: examples/generate_QG_time_series.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+# create dual iterable for the animated plot of truth / obs
+ts = zip(xx, yy_xx)
+
+# run over the dual time series
+for k, val in pb.progbar(list(enumerate(ts)), "Animating"):
+    if k % 2 == 0:
+        fig.suptitle("k: "+str(k))
+        setter1(val[0])
+        setter2(val[1])
```
Comment (patricknraanes) [inline examples/generate_QG_time_series.py:82]: In my testing the second axis does not show anything. Not sure why, maybe imshow doesn't work well with isolated pixels. Maybe try matshow or pcolormesh or something. However, I actually recommend you try `scatter` instead which gives you the opportunity to plot larger circles, like the envisat video from the tutorials, and then you won't have to mess around with `NaN`s. Be careful about the axes orientation and ordering. It often gives me headaches.

**Review by patricknraanes — CHANGES_REQUESTED** [review]

File: examples/generate_QG_time_series.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+from dapper.mods.QG.sakov2008 import obs_inds
+
+###########
+# Auxiliary plotting function
+###########
+
+def show(x0, psi=True, ax=None):
+    #
+    def psi_or_q(x):
+        return x if psi else compute_q(x)
```
Comment (patricknraanes) [inline examples/generate_QG_time_series.py:18]: `compute_q` not imported

File: examples/generate_QG_time_series.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+## of Sakov2008, save the double time series
+xx, yy = HMM.simulate()
+
+# save data for easy re-use and debugging
+np.savez("QG_truth_obs_time_series.npz", xx=xx, yy=yy)
+
+# these lines can be used in place of above steps after first run,
+# comment out above
+tmp = np.load("QG_truth_obs_time_series.npz")
+xx = tmp['xx'][1:,]
```
Comment (patricknraanes) [inline examples/generate_QG_time_series.py:48]: comma is unnecessary

File: examples/generate_QG_time_series.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+ax2.set_title(r'Noisy observations: $\mathcal{H}(\psi) + \epsilon$')
+
+setter1 = show(xx[0], psi=True, ax=ax1)
+setter2 = show(yy_xx[0], psi=True, ax=ax2)
+
+# create dual iterable for the animated plot of truth / obs
+ts = zip(xx, yy_xx)
+
+# run over the dual time series
+for k, val in pb.progbar(list(enumerate(ts)), "Animating"):
```
Comment (patricknraanes) [inline examples/generate_QG_time_series.py:78]:
```suggestion
for k, (xx, yy_xx) in pb.progbar(list(enumerate(ts)), "Animating"):
```

**Comment by cgrudz** [comment]
This gave me an unusual behavior too where the second pane didn't show anything until I resized the window -- upon doing so this did show the pixel observation plot however.  I'll take a whack at your suggestions and see if I can get to do about what I want.

File: examples/generate_QG_time_series.py (hunk 1 above)
Comment (patricknraanes) [reply examples/generate_QG_time_series.py:47]: Prettier, and has other minor advantages.
https://www.pythonforbeginners.com/files/with-statement-in-python

File: examples/generate_QG_time_series.py (hunk 3 above)
Comment (cgrudz) [reply examples/generate_QG_time_series.py:55]: This is a nice utility for sure!

File: examples/generate_QG_time_series.py (hunk 7 above)
Comment (cgrudz) [reply examples/generate_QG_time_series.py:78]: Sounds good, editing to include this.

File: examples/generate_QG_time_series.py (hunk 5 above)
Comment (cgrudz) [reply examples/generate_QG_time_series.py:18]: Cleaning up extra definitions, unused functions, this included.

File: examples/generate_QG_time_series.py (hunk 2 above)
Comment (cgrudz) [reply examples/generate_QG_time_series.py:1]: Word, I'm removing this as it is indeed just a time series demonstration of the dynamical and observation models.

File: examples/generate_QG_time_series.py (hunk 1 above)
Comment (cgrudz) [reply examples/generate_QG_time_series.py:47]: Ah, OK, I'll leave it with the close statement, but I have added an exception handler to open the saved data or generate it if it doesn't exist.

**Review by patricknraanes — APPROVED** [review]
Pretty nice

It'd be cool to try `scatter` too, but it already looks good!

Found the relevant doc for `with`
https://numpy.org/doc/stable/reference/generated/numpy.load.html

I did some linting so that it passes the checks.

Put the data in the configured data dir (enables avoiding putting in in your Dropbox, for example)

I suggest moving to the QG dir and some slight renaming. See if you agree.

Unfortunately I went about this slightly wrongly. My commits are on a new branch called pr95. I believe all your commits are still properly attributed though. Could you checkout that one and make a new PR on it (referencing this one in the comments)?

**Pull request closed without merging**